graphql_compiler/compiler/emit_sql.py

```diff
@@ -267,47 +349,59 @@ def _get_fold_outputs(self, fold_scope_location, join_descriptor, all_folded_out
+        if self._ended:
+            raise AssertionError(u'Cannot visit output vertices after end_fold has been called. '
+                                 u'Invalid state encountered during fold {}'.format(self))
         if self._output_vertex_alias is not None:
-            raise AssertionError('Cannot visit multiple output vertices in one fold.')
+            raise AssertionError(u'Cannot visit multiple output vertices in one fold. '
+                                 u'Invalid state encountered during fold {}'.format(self))
         self._output_vertex_alias = output_alias
         self._outputs = self._get_fold_outputs(fold_scope_location,
-                                               join_descriptor,
                                                all_folded_outputs)
 
     def visit_traversed_vertex(self, join_descriptor, from_table, to_table):
-        """Add a new join descriptor for every vertex traversed in the fold."""
-        self._join_info.append((join_descriptor, from_table, to_table))
+        """Add a new traversal descriptor for every vertex traversed in the fold."""
+        if self._ended:
+            raise AssertionError(u'Cannot visit traversed vertices after end_fold has been called.'
+                                 u'Invalid state encountered during fold {}'.format(self))
+        self._traversal_descriptors.append(SQLFoldTraversalDescriptor(join_descriptor,
+                                                                      from_table,
+                                                                      to_table))
 
     def end_fold(self, alias_generator, from_clause, outer_from_table):
         """Produce the final subquery and join it onto the rest of the query."""
+        if self._ended:
+            raise AssertionError(u'Cannot call end_fold more than once. '
+                                 u'Invalid state encountered during fold {}'.format(self))
         if self._output_vertex_alias is None:
-            raise AssertionError('No output vertex visited.')
-        if len(self._join_info) == 0:
-            raise AssertionError('No traversed vertices visited.')
+            raise AssertionError(u'No output vertex visited. '
+                                 u'Invalid state encountered during fold {}'.format(self))
+        if len(self._traversal_descriptors) == 0:
+            raise AssertionError(u'No traversed vertices visited. '
+                                 u'Invalid state encountered during fold {}'.format(self))
 
         # for now we only handle folds containing one traversal (i.e. join)
-        if len(self.join_info) > 1:
-            raise NotImplementedError('Folds containing multiple traversals are not implemented.')
-        edge, from_alias, to_alias = self.join_info.pop()
+        if len(self._traversal_descriptors) > 1:
+            raise NotImplementedError(u'Folds containing multiple traversals are not '
+                                      u'implemented {}.'.format(self))
```

Comment:
I think this error reads weird. The user will see `Folds containing multiple traversals are not implemented SQLFoldObject(...)` and it is kind of unclear to me what they're supposed to do with the information in `SQLFoldObject(...)`. Here the issue is that we haven't implemented something, not that the `SQLFoldObject` is being used incorrectly. Does having the `SQLFoldObject` information really benefit them more than seeing that there are multiple traversals inside a fold in the query itself?